### ISSUE TITLE
_$X not defined

### ISSUE THREAD

**Issue by jpohhhh** [d-markey/squadron_builder#19]
Using:
- squadron: ^6.0.3 in dependencies
- squadron_builder: ^6.0.2 in dev_dependencies

Input class: 
- https://gist.github.com/jpohhhh/230be0266fa7681cac00517fc2e66fc3

Command:
- dart run build_runner build

Output for tiktoken_service.worker.g.dart:
- https://gist.github.com/jpohhhh/b0d0f578113bc7c582bf6693ee7e8666 

Separate issue I ran into while trying to replicate generation in example. TL;DR: Can't get files generated in the example after deleting one.

I saw the example Data marshalling that goes through the generated $X. 
I attempted to verify that generation worked on my machine.
I deleted data_service_worker.g.dart.
I read README.md
I ran `dart run build_runner build` in squadron_builder checkout root.
data_service_worker.g.dart. was not regenerated.
I tried `cd example; dart run build_runner build` in /example.
The output was `Bad state: Unable to generate package graph, no `pubspec.yaml` found.`
I tried `cd example/misc, [build runner]`, same result.
Hewing closer to README.md, I tried doing the Hello World example.
I deleted hello_world.worker.g.dart, then `ran dart run build_runner build`. No file was generated.
I tried the same commands in example, example/hello_world, and got the same error re: pubspec.yaml not found.

**Comment by jpohhhh** [comment]
Also tried these, there is no difference in output:
1. adding `base` to the service class definition, a la https://github.com/d-markey/squadron_builder/blob/main/README.md
2. changing return type from `Future` to `FutureOr` and implementing marshallers manually a la https://github.com/d-markey/squadron/blob/main/README.md

code for 2: https://gist.github.com/jpohhhh/c455dd06f75f16cb75bebfe8401c8a04

**Comment by d-markey** [comment]
Hello, I can't reproduce your issue :-(

Regarding the test you did with `squadron_builder`checkout, you need to try this command line from the root dir of `squadron_builder`:

```
dart run build_runner build -d -c example
```

This is because the `build.yaml` only contains the default build settings, which build from `/lib`. The settings to build the examples are found in the `build.example.yaml` which is activated thanks to `-c example`.

**Comment by d-markey** [comment]
OK, thanks for the feedback. I've also come across occasional issues when generating / regenerating, that's why I generate examples with my `build_example.bat` script that starts with `dart run build_runner clean`...

I actually never got into the `WorkerGenerator.generate() is not reentrant!` bit, I'm curious as to why this happened. The "shared" type manager was clearly a code smell so I added this check just in case... well, you caught it! And you nailed it too, thanks for that. I'll integrate your change in the next version if you don't mind.

**Comment by jpohhhh** [comment]
I don't mind at all, please feel free: glad to hear it was sensical, felt a
bit hacky and I didn't know what I was doing, but sometimes, the best fixes
are like that :)

**Comment by d-markey** [comment]
BTW, I saw the comment "doesn't sound right to me that squadron methods are async".

This is simply because the Worker implements the service as an interface. And since communication with workers is async, the service must expose async methods. `FutureOr` is merely a trick to enable sync implementation at the service level, while allowing overriding the method to return a `Future` in the workers :) Another benefit of `FutureOr` is it silents the linter.

**Comment by jpohhhh** [comment]
Hah, thanks for noticing: I dropped a word there, should say are *not* async. 

tl;dr Dart Code Metrics flags it as an unnecessary future, but, it's necessary. (DCM is **awesome**, highly recommended, but there's some false positive.)

Also, I was thinking on the reentrant log. I still believe there was/is some local state issue on my end. But it hit me I might be challenging build_runner's async behavior, I have 12(!) services now with a ton of code/deps involved. Squadron makes it possible for me to run a full AI search engine stack in the browser, among other things.

**Comment by d-markey** [comment]
Update published, thanks again for your contribution.

Your fix doesn't feel like a hack to me, as I'm using the same approach to consolidate multiple outputs from a build event. The solution was right under my nose!

**Comment by d-markey** [comment]
BTW are you using `build_runner watch`? The concurrency issues might originate from this?
